packages/babel-parser/src/types.js

```diff
@@ -537,6 +537,14 @@ export type MemberExpression = NodeBase & {
   computed: boolean,
 };
 
+export type EventualMemberExpression = NodeBase & {
```

Comment:
w.l.o.g. I will take `EventualMemberExpression` as an example, likewise `EventualCallExpression`.

I am hesitant about the extra node type `EventualMemberExpression` and likely-in-the-future `EventualOptionalMemberExpression` / `OptionalEventualMemberExpression` given that optional chaining has reached stage 4.

The `wavy-dot` proposal introduces infix operators `~.` similar to optional chaining. I understand that it may seem reasonable to add a new node type, which is what we did in `OptionalMemberExpression`. Essentially `OptionalMemberExpression` is a `MemberExpression`. The introduction of the new node type `OptionalMemberExpression` is due to 1) the new optional semantic and 2) the short-circuit behaviour. The latter requires that a new node type has to be introduced because we need to know when the base is nullish, we should skip which parts of the whole optional chains.

`EventualMemberExpression` is also a `MemberExpression`. Similar to `OptionalMemberExpression`, it introduces new eventual semantics. Besides that I don't see there are extra behaviours which we should track on the AST level. Therefore I propose to add an `eventual: boolean` property to the existing `MemberExpression`

```js
interface MemberExpression <: Expression, Pattern {
  type: "MemberExpression";
  object: Expression | Super;
  property: Expression;
  computed: boolean;
  eventual: boolean;
}
```

Since the eventual semantic is additive, we can also add `eventual: boolean` once the proposal supports optional chaining.

Reply:
I'll remove for `EventualCallExpression` and `EventualMemberExpression`.  I think we still need `EventualMemberCallExpression`, which I would rename to `EventualSendExpression`.  I'll wait on that until you post your updated suggestion.

Reply:
Also, just a note:

`.?~` is sensible (only eventual-send if the lhs is not nullish), but `.~?` is not sensible (since eventual-send always returns a Promise, it cannot be nullish).

Reply:
> Updates: As @nicolo-ribaudo points out `(a~.b)()` is `a~.b()` is not equivalent. So adding `eventual: boolean` is not enough. I will post an updated suggestion tomorrow. I am still negative on adding new node types.

@JLHwung have you had a chance to think about this?  I definitely want to implement the PR in the correct way, but am blocked until I get some design advice.

Thanks!

Reply:
@michaelfig My approach is now based on the `ChainingExpression` approach proposed by ESLint team: See https://gist.github.com/JLHwung/6ec08a87e4da88874c50788c37d6fdf4 for detailed examples.

Note that this approach is tremendously different from current babel's approach and we will discuss if we will go for that path in the upcoming babel meetings. Before any decision is made, I would suggest we hold on a bit until more feedback from the AST shape of `wavy-dot`.

Reply:
[From the above, I found estree/estree#204 which is long, but seems to be giving some traction to https://gist.github.com/mysticatea/5453e721892b9c5a621e445f5fcdf62f]

Reply:
> @michaelfig My approach is now based on the `ChainingExpression` approach proposed by ESLint team: See https://gist.github.com/JLHwung/6ec08a87e4da88874c50788c37d6fdf4 for detailed examples.
> 
> Note that this approach is tremendously different from current babel's approach and we will discuss if we will go for that path in the upcoming babel meetings. Before any decision is made, I would suggest we hold on a bit until more feedback from the AST shape of `wavy-dot`.

@JLHwung, has this discussion settled?  If so, what does it mean for the kind of AST shape I should target?